File: applications/luci-app-unbound/luasrc/model/cbi/unbound/configure.lua
-- Copyright 2008 Steven Barth <steven@midlink.org>
-- Copyright 2016 Eric Luehrsen <ericluehrsen@gmail.com>
-- Copyright 2016 Dan Luedtke <mail@danrl.com>
-- Licensed to the public under the Apache License 2.0.

local m1, s1
local ena, mcf, lci, lsv
local rlh, rpv, vld, nvd, eds, prt, tlm
local ctl, dlk, dom, dty, lfq, wfq, exa
local dp6, d64, pfx, qry, qrs
local pro, rsc, rsn, ag2, stt
local tgr, ifc, wfc
local rpn, din, ath

local ut = require "luci.util"
local sy = require "luci.sys"
local ht = require "luci.http"
local ds = require "luci.dispatcher"
local ucl = luci.model.uci.cursor()
local valman = ucl:get_first("unbound", "unbound", "manual_conf")
local dhcplk = ucl:get_first("unbound", "unbound", "dhcp_link")
local lstrig = ucl:get_first("dhcp", "odhcpd", "leasetrigger") or "undefined"

m1 = Map("unbound")
s1 = m1:section(TypedSection, "unbound", translate("Recursive DNS"),
    translatef("Unbound <a href=\"%s\" target=\"_blank\">(NLnet Labs)</a>"
    .. " is a validating, recursive, and caching DNS resolver"
    .. " <a href=\"%s\" target=\"_blank\">(help)</a>.",
    "https://www.unbound.net/",
    "https://github.com/openwrt/packages/blob/master/net/unbound/files/README.md"))

s1.addremove = false
s1.anonymous = true

if (valman == "0") and (dhcplk == "odhcpd") and (lstrig ~= "/usr/lib/unbound/odhcpd.sh") then
    m1.message = translatef( "Note: local DNS is configured to look at odhpcd, "
    .. "but odhpcd UCI lease trigger is incorrectly set: ")
    .. "dhcp.odhcpd.leasetrigger='" .. lstrig .. "'"
end

--LuCI, Unbound, or Not
s1:tab("basic", translate("Basic"))


if (valman == "0") then
    -- Not in manual configuration mode; show UCI
    s1:tab("advanced", translate("Advanced"))
    s1:tab("DHCP", translate("DHCP"))
    s1:tab("resource", translate("Resource"))
end


--Basic Tab, unconditional pieces
ena = s1:taboption("basic", Flag, "enabled", translate("Enable Unbound"),
    translate("Enable the initialization scripts for Unbound"))
ena.rmempty = false

mcf = s1:taboption("basic", Flag, "manual_conf", translate("Manual Conf"),
    translate("Skip UCI and use /etc/unbound/unbound.conf"))
mcf.rmempty = false


if (valman == "0") then
    -- Not in manual configuration mode; show UCI
    --Basic Tab
    lsv = s1:taboption("basic", Flag, "localservice",
        translate("Local Service"),
        translate("Accept queries only from local subnets"))
    lsv.rmempty = false

    vld = s1:taboption("basic", Flag, "validator",
        translate("Enable DNSSEC"),
        translate("Enable the DNSSEC validator module"))
    vld.rmempty = false

    nvd = s1:taboption("basic", Flag, "validator_ntp",
        translate("DNSSEC NTP Fix"),
        translate("Break the loop where DNSSEC needs NTP and NTP needs DNS"))
    nvd.optional = true
    nvd.default = true
    nvd:depends("validator", true)

    prt = s1:taboption("basic", Value, "listen_port",
        translate("Listening Port"),
        translate("Choose Unbounds listening port"))
    prt.datatype = "port"
    prt.placeholder = "53"

    --Avanced Tab
    rlh = s1:taboption("advanced", Flag, "rebind_localhost",
        translate("Filter Localhost Rebind"),
        translate("Protect against upstream response of 127.0.0.0/8"))
    rlh.rmempty = false

    rpv = s1:taboption("advanced", ListValue, "rebind_protection",
        translate("Filter Private Rebind"),
        translate("Protect against upstream responses within local subnets"))
    rpv:value("0", translate("No Filter"))
    rpv:value("1", translate("Filter Private Address"))
    rpv:value("2", translate("Filter Entire Subnet"))
    rpv.rmempty = false

    d64 = s1:taboption("advanced", Flag, "dns64", translate("Enable DNS64"),
        translate("Enable the DNS64 module"))
    d64.rmempty = false

    pfx = s1:taboption("advanced", Value, "dns64_prefix",
        translate("DNS64 Prefix"),
        translate("Prefix for generated DNS64 addresses"))
    pfx.datatype = "ip6addr"
    pfx.placeholder = "64:ff9b::/96"
    pfx.optional = true
    pfx:depends("dns64", true)

    din = s1:taboption("advanced", DynamicList, "domain_insecure",
        translate("Domain Insecure"),
        translate("List domains to bypass checks of DNSSEC"))
    din:depends("validator", true)

    ag2 = s1:taboption("advanced", Value, "root_age",
        translate("Root DSKEY Age"),
        translate("Limit days between RFC5011 copies to reduce flash writes"))
    ag2.datatype = "and(uinteger,min(1),max(99))"
    ag2:value("3", "3")
    ag2:value("9", "9 ("..translate("default")..")")
    ag2:value("12", "12")
    ag2:value("24", "24")
    ag2:value("99", "99 ("..translate("never")..")")

    ifc = s1:taboption("advanced", Value, "iface_lan",
        translate("LAN Networks"),
        translate("Networks to consider LAN (served) beyond those served by DHCP"))
    ifc.template = "cbi/network_netlist"
    ifc.widget = "checkbox"
    ifc.rmempty = true
    ifc.cast = "string"
    ifc.nocreate = true

    wfc = s1:taboption("advanced", Value, "iface_wan",
        translate("WAN Networks"),
        translate("Networks to consider WAN (unserved)"))
    wfc.template = "cbi/network_netlist"
    wfc.widget = "checkbox"
    wfc.rmempty = true
    wfc.cast = "string"
    wfc.nocreate = true

    tgr = s1:taboption("advanced", Value, "iface_trig",
        translate("Trigger Networks"),
        translate("Networks that may trigger Unbound to reload (avoid wan6)"))
    tgr.template = "cbi/network_netlist"
    tgr.widget = "checkbox"
    tgr.rmempty = true
    tgr.cast = "string"
    tgr.nocreate = true

    --DHCP Tab
    dlk = s1:taboption("DHCP", ListValue, "dhcp_link",
        translate("DHCP Link"),
        translate("Link to supported programs to load DHCP into DNS"))
    dlk:value("none", translate("(none)"))
    dlk:value("dnsmasq", "dnsmasq")
    dlk:value("odhcpd", "odhcpd")
    dlk.rmempty = false

    dp6 = s1:taboption("DHCP", Flag, "dhcp4_slaac6",
        translate("DHCPv4 to SLAAC"),
        translate("Use DHCPv4 MAC to discover IP6 hosts SLAAC (EUI64)"))
    dp6.optional = true
    dp6:depends("dhcp_link", "odhcpd")

    dom = s1:taboption("DHCP", Value, "domain",
        translate("Local Domain"),
        translate("Domain suffix for this router and DHCP clients"))
    dom.placeholder = "lan"
    dom.optional = true

    dty = s1:taboption("DHCP", ListValue, "domain_type",
        translate("Local Domain Type"),
        translate("How to treat queries of this local domain"))
    dty.optional = true
    dty:value("deny", translate("Denied (nxdomain)"))
    dty:value("refuse", translate("Refused"))
    dty:value("static", translate("Static (local only)"))
    dty:value("transparent", translate("Transparent (local/global)"))
    dty:depends("dhcp_link", "none")
    dty:depends("dhcp_link", "odhcpd")

    lfq = s1:taboption("DHCP", ListValue, "add_local_fqdn",
        translate("LAN DNS"),
        translate("How to enter the LAN or local network router in DNS"))
    lfq.optional = true
    lfq:value("0", translate("No Entry"))
    lfq:value("1", translate("Hostname, Primary Address"))
    lfq:value("2", translate("Hostname, All Addresses"))
    lfq:value("3", translate("Host FQDN, All Addresses"))
    lfq:value("4", translate("Interface FQDN, All Addresses"))
    lfq:depends("dhcp_link", "none")
    lfq:depends("dhcp_link", "odhcpd")

    wfq = s1:taboption("DHCP", ListValue, "add_wan_fqdn",
        translate("WAN DNS"),
        translate("Override the WAN side router entry in DNS"))
    wfq.optional = true
    wfq:value("0", translate("Use Upstream"))
    wfq:value("1", translate("Hostname, Primary Address"))
    wfq:value("2", translate("Hostname, All Addresses"))
    wfq:value("3", translate("Host FQDN, All Addresses"))
    wfq:value("4", translate("Interface FQDN, All Addresses"))
    wfq:depends("dhcp_link", "none")
    wfq:depends("dhcp_link", "odhcpd")

    exa = s1:taboption("DHCP", ListValue, "add_extra_dns",
        translate("Extra DNS"),
        translate("Use extra DNS entries found in /etc/config/dhcp"))
    exa.optional = true
    exa:value("0", translate("Ignore"))
    exa:value("1", translate("Host Records"))
    exa:value("2", translate("Host/MX/SRV RR"))
    exa:value("3", translate("Host/MX/SRV/CNAME RR"))
    exa:depends("dhcp_link", "none")
    exa:depends("dhcp_link", "odhcpd")

    --TODO: dnsmasq needs to not reference resolve-file and get off port 53.

    --Resource Tuning Tab
    ctl = s1:taboption("resource", ListValue, "unbound_control",
        translate("Unbound Control App"),
        translate("Enable access for unbound-control"))
    ctl.rmempty = false
    ctl:value("0", translate("No Remote Control"))
    ctl:value("1", translate("Local Host, No Encryption"))
    ctl:value("2", translate("Local Host, Encrypted"))
    ctl:value("3", translate("Local Subnet, Encrypted"))
    ctl:value("4", translate("Local Subnet, Static Encryption"))

    pro = s1:taboption("resource", ListValue, "protocol",
        translate("Recursion Protocol"),
        translate("Choose the IP versions used upstream and downstream"))
    pro:value("default", translate("Default"))
    pro:value("ip4_only", translate("IP4 Only"))
    pro:value("ip6_local", translate("IP4 All and IP6 Local"))
    pro:value("ip6_only", translate("IP6 Only*"))
    pro:value("ip6_prefer", translate("IP6 Preferred"))
    pro:value("mixed", translate("IP4 and IP6"))
    pro.rmempty = false

    rsc = s1:taboption("resource", ListValue, "resource",
        translate("Memory Resource"),
        translate("Use menu System/Processes to observe any memory growth"))
    rsc:value("default", translate("Default"))
    rsc:value("tiny", translate("Tiny"))
    rsc:value("small", translate("Small"))
    rsc:value("medium", translate("Medium"))
    rsc:value("large", translate("Large"))
    rsc.rmempty = false

    rsn = s1:taboption("resource", ListValue, "recursion",
        translate("Recursion Strength"),
        translate("Recursion activity affects memory growth and CPU load"))
    rsn:value("default", translate("Default"))
    rsn:value("passive", translate("Passive"))
    rsn:value("aggressive", translate("Aggressive"))
    rsn.rmempty = false

    qry = s1:taboption("resource", Flag, "query_minimize",
        translate("Query Minimize"),
        translate("Break down query components for limited added privacy"))
    qry.optional = true
    qry:depends("recursion", "passive")
    qry:depends("recursion", "aggressive")

    qrs = s1:taboption("resource", Flag, "query_min_strict",
        translate("Strict Minimize"),
        translate("Strict version of 'query minimize' but it can break DNS"))
    qrs.optional = true
    qrs:depends("query_minimize", true)

    eds = s1:taboption("resource", Value, "edns_size",
        translate("EDNS Size"),
        translate("Limit extended DNS packet size"))
    eds.datatype = "and(uinteger,min(512),max(4096))"
    eds.placeholder = "1280"

    tlm = s1:taboption("resource", Value, "ttl_min",
        translate("TTL Minimum"),
        translate("Prevent excessively short cache periods"))
    tlm.datatype = "and(uinteger,min(0),max(1200))"
    tlm.placeholder = "120"

    rtt = s1:taboption("resource", Value, "rate_limit",
        translate("Query Rate Limit"),
        translate("Prevent client query overload; zero is off"))
    rtt.datatype = "and(uinteger,min(0),max(5000))"
    rtt.placeholder = "0"

    stt = s1:taboption("resource", Flag, "extended_stats",
        translate("Extended Statistics"),
        translate("Extended statistics are printed from unbound-control"))
    stt.rmempty = false

else
    ag2 = s1:taboption("basic", Value, "root_age",
        translate("Root DSKEY Age"),
        translate("Limit days between RFC5011 copies to reduce flash writes"))
    ag2.datatype = "and(uinteger,min(1),max(99))"
    ag2:value("3", "3")
    ag2:value("9", "9 ("..translate("default")..")")
    ag2:value("12", "12")
    ag2:value("24", "24")
    ag2:value("99", "99 ("..translate("never")..")")

    tgr = s1:taboption("basic", Value, "trigger_interface",
        translate("Trigger Networks"),
        translate("Networks that may trigger Unbound to reload (avoid wan6)"))
    tgr.template = "cbi/network_netlist"
    tgr.widget = "checkbox"
    tgr.rmempty = true
    tgr.cast = "string"
    tgr.nocreate = true
end


function ena.cfgvalue(self, section)
    return sy.init.enabled("unbound") and self.enabled or self.disabled
end


function ena.write(self, section, value)
    if (value == "1") then
        sy.init.enable("unbound")
        sy.call("/etc/init.d/unbound start >/dev/null 2>&1")

    else
        sy.call("/etc/init.d/unbound stop >/dev/null 2>&1")
        sy.init.disable("unbound")
    end


    return Flag.write(self, section, value)
end


function m1.on_commit(self)
    if sy.init.enabled("unbound") then
        -- Restart Unbound with configuration
        sy.call("/etc/init.d/unbound restart >/dev/null 2>&1")

    else
        sy.call("/etc/init.d/unbound stop >/dev/null 2>&1")
    end
end


function m1.on_apply(self)
    -- reload the page because some options hide
    ht.redirect(ds.build_url("admin", "services", "unbound", "configure"))
end


return m1

File: applications/luci-app-unbound/htdocs/luci-static/resources/view/unbound/configure.js
'use strict';
'require fs';
'require form';
'require rpc';
'require uci';
'require ui';
'require view';
'require tools.widgets as widgets';

const UBNET = 'https://nlnetlabs.nl/projects/unbound/about/';
const README = 'https://github.com/openwrt/packages/blob/master/net/unbound/files/README.md';

const callRcInit = rpc.declare({
	object: 'rc',
	method: 'init',
	params: [ 'name', 'action' ],
	expect: { result: false }
});

return view.extend({
	load() {
		return Promise.all([
			uci.load('unbound'),
			uci.load('dhcp'),
		]);
	},

	render([unboundCfg, dhcpCfg]) {
		const m = new form.Map('unbound', _('Recursive DNS'),
			_("Unbound <a href=\"%s\" target=\"_blank\">(NLnet Labs)</a> is a validating, recursive, and caching DNS resolver <a href=\"%s\" target=\"_blank\">(help)</a>.".format(UBNET, README)));

		const s = m.section(form.NamedSection, 'ub_main', 'unbound');
		s.anonymous = true;
		s.addremove = false;

		const manual = uci.get('unbound', 'ub_main', 'manual_conf');
		const dhcpLink = uci.get('unbound', 'ub_main', 'dhcp_link');
		const leasetrig = uci.get('dhcp', 'odhcpd', 'leasetrigger') || 'undefined';

		if (manual === '0' && dhcpLink === 'odhcpd' && leasetrig !== '/usr/lib/unbound/odhcpd.sh') {
			ui.addTimeLimitedNotification(null, E('p', _('Note: local DNS is configured to look at odhpcd, but odhpcd UCI lease trigger is incorrectly set:') +
				"dhcp.odhcpd.leasetrigger='" + leasetrig + "'"), 15000, 'note');
		}

		// Tabs
		s.tab('basic', _('Basic'));
		if (manual === '0') {
			s.tab('advanced', _('Advanced'));
			s.tab('dhcp', _('DHCP'));
			s.tab('resource', _('Resource'));
		}

		// Basic tab items
		const ena = s.taboption('basic', form.Flag, 'enabled', _('Enable Unbound'),
			_('Enable the initialization scripts for Unbound'));
		ena.rmempty = false;

		const mcf = s.taboption('basic', form.Flag, 'manual_conf', _('Manual Conf'),
			_('Skip UCI and use /etc/unbound/unbound.conf'));
		mcf.rmempty = false;

		if (manual === '0') {
			s.taboption('basic', form.Flag, 'localservice', _('Local Service'),
				_('Accept queries only from local subnets')).rmempty = false;

			s.taboption('basic', form.Flag, 'validator', _('Enable DNSSEC'),
				_('Enable the DNSSEC validator module')).rmempty = false;

			const nvd = s.taboption('basic', form.Flag, 'validator_ntp', _('DNSSEC NTP Fix'),
				_('Break the loop where DNSSEC needs NTP and NTP needs DNS'));
			nvd.optional = true;
			nvd.default = '1';
			nvd.depends('validator', '1');

			const prt = s.taboption('basic', form.Value, 'listen_port', _('Listening Port'),
				_('Choose Unbounds listening port'));
			prt.datatype = 'port';
			prt.placeholder = '53';
		} else {
			const ag2b = s.taboption('basic', form.Value, 'root_age', _('Root DSKEY Age'),
				_('Limit days between RFC5011 copies to reduce flash writes'));
			ag2b.datatype = 'and(uinteger,min(1),max(99))';
			ag2b.value('3', '3');
			ag2b.value('9', '9 (' + _('default') + ')');
			ag2b.value('12', '12');
			ag2b.value('24', '24');
			ag2b.value('99', '99 (' + _('never') + ')');

			const tgrb = s.taboption('basic', widgets.NetworkSelect, 'trigger_interface', _('Trigger Networks'),
				_('Networks that may trigger Unbound to reload (avoid wan6)'));
			tgrb.multiple = true;
			tgrb.rmempty = true;
			tgrb.nocreate = true;
		}

		// Advanced, DHCP & Resource tabs only if manual===0
		if (manual === '0') {
			const rlh = s.taboption('advanced', form.Flag, 'rebind_localhost', _('Filter Localhost Rebind'),
				_('Protect against upstream response of 127.0.0.0/8'));
			rlh.rmempty = false;

			const rpv = s.taboption('advanced', form.ListValue, 'rebind_protection', _('Filter Private Rebind'),
				_('Protect against upstream responses within local subnets'));
			rpv.rmempty = false;
			rpv.value('0', _('No Filter'));
			rpv.value('1', _('Filter Private Address'));
			rpv.value('2', _('Filter Entire Subnet'));

			const d64 = s.taboption('advanced', form.Flag, 'dns64', _('Enable DNS64'),
				_('Enable the DNS64 module'));
			d64.rmempty = false;

			const pfx = s.taboption('advanced', form.Value, 'dns64_prefix', _('DNS64 Prefix'),
				_('Prefix for generated DNS64 addresses'));
			pfx.datatype = 'ip6addr';
			pfx.placeholder = '64:ff9b::/96';
			pfx.optional = true;
			pfx.depends('dns64', '1');

			const exga = s.taboption('advanced', form.Flag, 'exclude_ipv6_ga', _('Exclude IPv6 GA'));

			const din = s.taboption('advanced', form.DynamicList, 'domain_insecure', _('Domain Insecure'),
				_('List domains to bypass checks of DNSSEC'));
			din.depends('validator', '1');

			const ag2 = s.taboption('advanced', form.Value, 'root_age', _('Root DSKEY Age'));
			ag2.datatype = 'and(uinteger,min(1),max(99))';
			ag2.value('3', '3');
			ag2.value('9', '9 (' + _('default') + ')');
			ag2.value('12', '12');
			ag2.value('24', '24');
			ag2.value('99', '99 (' + _('never') + ')');

			const ifc = s.taboption('advanced', widgets.NetworkSelect, 'iface_lan', _('LAN Networks'), 
					_('Networks to consider LAN (served) beyond those served by DHCP'));
			ifc.multiple = true;
			ifc.rmempty = true;
			ifc.nocreate = true;
			const wfc = s.taboption('advanced', widgets.NetworkSelect, 'iface_wan', _('WAN Networks'), 
					_('Networks to consider WAN (unserved)'));
			wfc.multiple = true;
			wfc.rmempty = true;
			wfc.nocreate = true;
			const tgr = s.taboption('advanced', widgets.NetworkSelect, 'iface_trig', _('Trigger Networks'), 
					_('Networks that may trigger Unbound to reload (avoid wan6)'));
			tgr.multiple = true;
			tgr.rmempty = true;
			tgr.nocreate = true;

			const bindhide = s.taboption('advanced', form.Flag, 'hide_binddata', _('Refuse possible attack queries'));

			const verb = s.taboption('advanced', form.ListValue, 'verbosity', _('Verbosity'));
			verb.datatype = 'and(uinteger,min(0),max(5))';
			verb.value('0');
			verb.value('1');
			verb.value('2');
			verb.value('3');
			verb.value('4');
			verb.value('5');

			// DHCP tab
			const dlk = s.taboption('dhcp', form.ListValue, 'dhcp_link', _('DHCP Link'),
				_('Link to supported programs to load DHCP into DNS'));
			dlk.rmempty = false;
			['none','dnsmasq','odhcpd'].forEach(v => dlk.value(v, v === 'none' ? _('(none)') : v));

			const dp6 = s.taboption('dhcp', form.Flag, 'dhcp4_slaac6', _('DHCPv4 to SLAAC'),
				_('Use DHCPv4 MAC to discover IP6 hosts SLAAC (EUI64)'));
			dp6.optional = true;
			dp6.depends('dhcp_link', 'odhcpd');

			const dom = s.taboption('dhcp', form.Value, 'domain', _('Local Domain'),
				_('Domain suffix for this router and DHCP clients'));
			dom.placeholder = 'lan';
			dom.optional = true;

			const dty = s.taboption('dhcp', form.ListValue, 'domain_type', _('Local Domain Type'),
				_('How to treat queries of this local domain'));
			dty.optional = true;
			dty.value('deny', _('Denied (nxdomain)'));
			dty.value('refuse', _('Refused'));
			dty.value('static', _('Static (local only)'));
			dty.value('transparent', _('Transparent (local/global)'));
			dty.depends('dhcp_link', 'none');
			dty.depends('dhcp_link', 'odhcpd');

			const addDnsOpts = [
				['add_local_fqdn', _('LAN DNS'), _('How to enter the LAN or local network router in DNS')],
				['add_wan_fqdn', _('WAN DNS'), _('Override the WAN side router entry in DNS')],
			];
			const addDnsValues = [
				['1', _('Hostname, Primary Address')],
				['2', _('Hostname, All Addresses')],
				['3', _('Host FQDN, All Addresses')],
				['4', _('Interface FQDN, All Addresses')]
			];

			addDnsOpts.forEach(([optName,label]) => {
				const opt = s.taboption('dhcp', form.ListValue, optName, label, '');
				opt.optional = true;
				opt.value('0', (optName == 'add_local_fqdn') ? _('No Entry') : _('Use Upstream'));
				addDnsValues.forEach(([val,text]) => opt.value(val, text));
				opt.depends('dhcp_link', 'none');
				opt.depends('dhcp_link', 'odhcpd');
			});

			const exa = s.taboption('dhcp', form.ListValue, 'add_extra_dns', _('Extra DNS'),
				_('Use extra DNS entries found in /etc/config/dhcp'));
			exa.optional = true;
			exa.value('0', _('Ignore'));
			exa.value('1', _('Host Records'));
			exa.value('2', _('Host/MX/SRV RR'));
			exa.value('3', _('Host/MX/SRV/CNAME RR'));

			// Resource tab
			const ctl = s.taboption('resource', form.ListValue, 'unbound_control', _('Unbound Control App'),
				_('Enable access for unbound-control'));
			ctl.value('0', _('No Remote Control'));
			ctl.value('1', _('Local Host, No Encryption'));
			ctl.value('2', _('Local Host, Encrypted'));
			ctl.value('3', _('Local Subnet, Encrypted'));
			ctl.value('4', _('Local Subnet, Static Encryption'));
			ctl.rmempty = false;

			const pro = s.taboption('resource', form.ListValue, 'protocol', _('Recursion Protocol'),
				_('Choose the IP versions used upstream and downstream'));
			pro.value('default', _('Default'));
			pro.value('ip4_only', _('IP4 Only'));
			pro.value('ip6_local', _('IP4 All and IP6 Local'));
			pro.value('ip6_only', _('IP6 Only*'));
			pro.value('ip6_prefer', _('IP6 Preferred'));
			pro.value('mixed', _('IP4 and IP6'));
			pro.rmempty = false;

			const rsc = s.taboption('resource', form.ListValue, 'resource', _('Memory Resource'),
				_('Use menu System/Processes to observe any memory growth'));
			rsc.value('default', _('Default'));
			rsc.value('tiny', _('Tiny'));
			rsc.value('small', _('Small'));
			rsc.value('medium', _('Medium'));
			rsc.value('large', _('Large'));
			rsc.rmempty = false;

			const rsn = s.taboption('resource', form.ListValue, 'recursion', _('Recursion Strength'),
				_('Recursion activity affects memory growth and CPU load'));
			rsn.value('default', _('Default'));
			rsn.value('passive', _('Passive'));
			rsn.value('aggressive', _('Aggressive'));
			rsn.rmempty = false;

			const qry = s.taboption('resource', form.Flag, 'query_minimize', _('Query Minimize'),
				_('Break down query components for limited added privacy'));
			qry.optional = true;
			qry.depends('recursion', 'passive');
			qry.depends('recursion', 'aggressive');

			const qrs = s.taboption('resource', form.Flag, 'query_min_strict', _('Strict Minimize'),
				_("Strict version of 'query minimize' but it can break DNS"));
			qrs.taboptional = true;
			qrs.depends('query_minimize', '1');

			const eds = s.taboption('resource', form.Value, 'edns_size', _('EDNS Size'),
				_('Limit extended DNS packet size'));
			eds.datatype = 'and(uinteger,min(512),max(4096))';
			eds.placeholder = '1280';

			const tlm = s.taboption('resource', form.Value, 'ttl_min', _('TTL Minimum'),
				_('Prevent excessively short cache periods'));
			tlm.datatype = 'and(uinteger,min(0),max(1200))';
			tlm.placeholder = '120';

			const tlnm = s.taboption('resource', form.Value, 'ttl_neg_max', _('TTL Neg Max'));
			tlm.datatype = 'and(uinteger,min(0),max(1200))';
			tlm.placeholder = '1000';

			const rtt = s.taboption('resource', form.Value, 'rate_limit', _('Query Rate Limit'),
				_('Prevent client query overload; zero is off'));
			rtt.datatype = 'and(uinteger,min(0),max(5000))';
			rtt.placeholder = '0';

			const stt = s.taboption('resource', form.Flag, 'extended_stats', _('Extended Statistics'),
				_('Extended statistics are printed from unbound-control'));
			stt.rmempty = false;
		}

		return m.render();
	},

	handleSaveApply(ev, mode) {
		const enabled = uci.get('unbound', 'ub_main', 'enabled');
		const cmd = enabled ? 'reload' : 'stop';

		const Fn = L.bind(() => {
			callRcInit('unbound', cmd);
			document.removeEventListener('uci-applied', Fn);
		});
		document.addEventListener('uci-applied', Fn);
		this.super('handleSaveApply', [ev, mode]);
	}
});
